### PR TITLE
Support arbitrary client metadata, avoid linear seach in sendCap.

### PR DESCRIPTION
**Pull request by zenhack**
This solves an issue I discovered where sendCap() does a linear search
on the exports table for each sent capability.

After a long discussion with Louis over matrix, we concluded that adding
a metadata object for clients was the best solution; there's a general
design awkwardness right now where the rpc system needs to be able to
inspect the client in various ways, but Client is defined in the main
package and tries to encapsulate its implementation details. To some
degree we are able to use Brand(), but this gets awkward quickly,
because:

1. ClientHook is an openly implementable interface, so we don't know
   all the posibilities in the rpc code.
2. As I've been working on supporting receiverAnswer, I've found
   pinning down all the implementations I defined in the library that
   need to be worried about is challenging.
3. Some of these implementations aren't even exported, or just return
   nil from Brand().

The Metadata object lets the rpc system manage the bookkeeping
information independently of the internals of a client.

I expect to use this for more in the future, but for now this lets
us avoid the linear search by just storing the export id for each
connection in the client's metadata.

### CODE REVIEW

File: capability.go
Comment (zenhack) [inline capability.go:466]: Note that I'm not 100% sure these semantics are actually what is desired, or at least it seems like we might want to still be able to access the un-resolved version of this. `State()` is supposed to return a snapshot, so perhaps we also want `OriginalState()` which would recover pre-resolution state (in which case if `OrigState().IsPromise == false` then `State() == OrigState()`)?

...I feel like I don't totally understand the broader use of this, and I kindof expect whatever decisions we make here we'll likely discover we need to tweak exactly how this works a bit before stabilizing.

File: capability.go
Comment (lthibault) [inline capability.go:142]: This more of a design/devex issue than anything else, but I feel that this should be handled by a constructor.  As it is, it adds visual noise to the code-base and raises the possibility of forgetting to call `init()`. 

In fact, I think `clientHook` should probably have a corresponding `newClientHook` function, but we don't have to address that in this PR.

File: capability.go
Comment (lthibault) [reply capability.go:466]: 👍  Copy that.

Perhaps consider adding a `TODO` comment?

File: rpc/export.go
Comment (lthibault) [inline rpc/export.go:140]: Shouldn't `next()` just return a correctly-typed `exportID`?  Again, visual noise.

**Review by lthibault — DISMISSED** [review]
Approving this because it looks correct.  Nevertheless, I feel moderately strongly that the two readability issues I've flagged should be addressed before merging.  Your call.

**Comment by lthibault** [comment]
Could [this check](https://github.com/capnproto/go-capnproto2/pull/208/files#diff-b3e7d47e460625492aa4e5f0d42b8195cd9179ab8f75def101604256ada2825bR115-R120) in `sendCap` be refactored to use `isLocalClient` from #212 ?

That would be a huge readability win.  The above snippet is quite jarring every time I look at it, despite the fact that the domain logic it represents is dead-simple.

File: rpc/export.go
Comment (zenhack) [reply rpc/export.go:140]: It was like that when I found it :P. But I think the reason is that `c.exportID` is of type `idgen`, which is used for generating other types of identifiers. Once generics lands, we could consider making idgen generic and changing this to be an `idgen(exportID)`. For now though, I think the code re-use outweights needing to cast.

File: capability.go
Comment (zenhack) [reply capability.go:466]: Done.

File: capability.go
Comment (zenhack) [reply capability.go:142]: The reason I did it this way is I wanted to have the `Metadata` live directly in the `clientHook`. There is `NewMetadata`, but it returns a pointer...

**Comment by zenhack** [comment]
> Could this check in sendCap be refactored to use isLocalClient from #212 ?

I assume you meant to link to a specific line? The link just points to the whole diff, so I'm not sure which check you mean.

File: capability.go
Comment (lthibault) [reply capability.go:142]: That's what I gathered, but couldn't we nevertheless return a concrete struct from the constructor?  It would copy the mutex, but there would no longer be any copies after first use.  Unless I'm misremembering, the Go toolchain should be perfectly happy with that.

**Comment by lthibault** [comment]
Looks like it was merged, so no longer appearing in the diff:  https://github.com/capnproto/go-capnproto2/blob/8c878e053da7da0aa3d7c46f9c60c1338efaa84a/rpc/export.go#L88-L93

However, I just saw that we actually use the concrete `*importClient`, so never mind.

File: capability.go
Comment (zenhack) [reply capability.go]: Ok, changed (I also tweaked the wording on the comment so it's clear it's ok to copy before first use, which addresses my main hangup).

I also merged in main and used syncutil.With in a couple places where it's appropriate.

**Review by lthibault — DISMISSED** [review]
Cheers, much nicer to look at :)

**Review by lthibault — APPROVED** [review]

**Pull request merged**
